pkg/storage/utils/decomposedfs/tree/propagator/async.go

```diff
@@ -0,0 +1,427 @@
+		doSleep := false // switch to not sleep on the first iteration
+		for {
+			if doSleep {
+				time.Sleep(5 * time.Minute)
+			}
+			doSleep = true
+			log.Debug().Msg("scanning for stale .processing dirs")
+
+			entries, err := filepath.Glob(changesDirPath + "/**/*")
+			if err != nil {
+				continue
+			}
+
+			for _, e := range entries {
+				changesDirPath := e
+				entry, err := os.Stat(changesDirPath)
+				if err != nil {
+					continue
+				}
+				// recover all dirs that haven't been propagated within 3 minutes
+				if !entry.IsDir() || time.Now().Before(entry.ModTime().Add(3*time.Minute)) {
+					continue
+				}
+
+				go func() {
+					if !strings.HasSuffix(changesDirPath, ".processing") {
+						// first rename the existing node dir
+						err = os.Rename(changesDirPath, changesDirPath+".processing")
+						if err != nil {
+							return
+						}
+						changesDirPath += ".processing"
+					}
+
+					log.Debug().Str("dir", changesDirPath).Msg("propagating stale .processing dir")
+					parts := strings.SplitN(entry.Name(), ":", 2)
+					now := time.Now()
+					_ = os.Chtimes(changesDirPath, now, now)
+					p.propagate(context.Background(), parts[0], strings.TrimSuffix(parts[1], ".processing"), true, *log)
+				}()
+			}
+		}
+	}()
+
+	return p
+}
+
+func (p AsyncPropagator) Propagate(ctx context.Context, n *node.Node, sizeDiff int64) error {
+	ctx, span := tracer.Start(ctx, "Propagate")
+	defer span.End()
+	log := appctx.GetLogger(ctx).With().
+		Str("method", "async.Propagate").
+		Str("spaceid", n.SpaceID).
+		Str("nodeid", n.ID).
+		Str("parentid", n.ParentID).
+		Int64("sizeDiff", sizeDiff).
+		Logger()
+
+	if !p.treeTimeAccounting && (!p.treeSizeAccounting || sizeDiff == 0) {
+		// no propagation enabled
+		log.Debug().Msg("propagation disabled or nothing to propagate")
+		return nil
+	}
+
+	// add a change to the parent node
+	c := Change{
+		// use a sync time and don't rely on the mtime of the current node, as the stat might not change when a rename happened too quickly
+		SyncTime: time.Now().UTC(),
+		SizeDiff: sizeDiff,
+	}
+	go p.queuePropagation(ctx, n.SpaceID, n.ParentID, c, log)
+
+	return nil
+}
+
+func (p AsyncPropagator) queuePropagation(ctx context.Context, spaceID, nodeID string, change Change, log zerolog.Logger) {
+	// add a change to the parent node
+	changePath := p.changesPath(spaceID, nodeID, uuid.New().String()+".mpk")
+
+	data, err := msgpack.Marshal(change)
+	if err != nil {
+		log.Error().Err(err).Msg("failed to marshal Change")
+		return
+	}
+
+	_, subspan := tracer.Start(ctx, "write changes file")
+	ready := false
+	triggerPropagation := false
+	_ = os.MkdirAll(filepath.Dir(filepath.Dir(changePath)), 0700)
+	err = os.Mkdir(filepath.Dir(changePath), 0700)
+	triggerPropagation = err == nil || os.IsExist(err) // only the first goroutine, which succeeds to create the directory, is supposed to actually trigger the propagation
+	for retries := 0; retries <= 500; retries++ {
```

Comment:
`500` retries seems absolutely random. Is there a reason for this?

Reply:
Well, the thing here is that we *really* want to write the change to disk and not lose this update, but failing to write is an expected case when changes directory has been moved by a propagation goroutine meanwhile and this could happen multiple times as well.

At the same time we can't keep trying indefinitely as we might pile up goroutines and run out of resources at some point.